Humanoid rig

0b5vr opened the pull request:
The second shot of #877 

### Description

- ✨ `VRMHumanoid`: Add methods to control normalized human bones. These methods provide access to normalized proxies of human bones.
    - `autoUpdateHumanBones` -- Whether it copies pose from normalizedHumanBones to rawHumanBones on `update()`. `true` by default.
    - `get normalizedRestPose` -- The rest pose of normalized human bones.
    - `get normalizedHumanBones` -- A map from `VRMHumanBoneName` to normalized human bones.
    - `getNormalizedAbsolutePose()` -- Retrieves the current absolute pose of normalized human bones.
    - `getNormalizedPose()` -- Retrieves the current pose of normalized human bones.
    - `setNormalizedPose()` -- Sets a pose of normalized human bones.
    - `resetNormalizedPose()` -- Reset the normalized human bones to its rest pose.
    - `getNormalizedBone()` -- Return a normalized `VRMHumanBone` bound to a specified `VRMHumanBoneName`.
    - `getNormalizedBoneNode()` -- Return a normalized bone as a `THREE.Object3D` bound to a specified `VRMHumanBoneName`.
- 💡🚨 Rename existing `VRMHumanoid` methods to distinguish them from their normalized variant. Previous names are still accessible but marked as deprecated.
    - `restPose` -> `rawRestPose`
    - `humanBones` -> `rawHumanBones` (the previous name is still valid)
    - `getAbsolutePose()` -> `getRawAbsolutePose()`
    - `getPose()` -> `getRawPose()`
    - `setPose()` -> `setRawPose()`
    - `resetPose()` -> `resetRawPose()`
    - `getBone()` -> `getRawBone()`
    - `getBoneNode()` -> `getRawBoneNode()`
- 💡 `VRMLoaderPlugin` / `VRMCoreLoaderPlugin` / `VRMHumanoidLoaderPlugin` : Add a new option `autoUpdateHumanBones`, which corresponds to the setting of `VRMHumanoid` mentioned above.
- ✨ New exported const `VRMHumanBonesList`: A list of `VRMHumanBoneName`, dependency aware.
- ✨ New exported const `VRMHumanBoneParentMap`: A map from `VRMHumanBoneName` to its parent `VRMHumanBoneName`.
- 🐛 `VRMHumanoidLoaderPluginOptions` was not exported; fixed this.
- 📄 Add a new example `humanoidAnimation`, which loads a [mixamo](https://www.mixamo.com/) animation and play on the VRM model

### Context

https://github.com/vrm-c/vrm-specification/issues/337

In VRM1.0, humanoid bones will no longer be reset to uniform transform.
The main purpose of this change is to make application developers able to control humanoid models with ease.

### Points need review

- [x] I have 0% confidence about the API and naming
    - I still have no idea how I should call this instead of `HumanoidRig` ...
    - We have settled the interface to a somewhat solid one. shoutouts to @ke456-png
- [ ] Are those interfaces usable?

### TODOs

- [ ] HumanoidRig probably should have missing intermediate bones

0b5vr committed:
feature (1.0, humanoid): add VRMHumanoid.transferBoneOrientations

0b5vr committed:
refactor (VRMHumanoid): add comments

0b5vr committed:
refactor (VRMHumanoid): add comments

0b5vr committed:
fix (VRMHumanoid): transferBoneOrientations should process bones from parent to children

0b5vr committed:
fix (VRMHumanoid): more explicit variable names

0b5vr committed:
fix (VRMHumanoid): transferBoneOrientations now receives hipsWorldQuat

0b5vr committed:
feature: trying VRMHumanoidRig

0b5vr committed:
change: remove the previous experimentation
- normalizeBoneOrientations
- transferBoneOrientations

0b5vr committed:
change: make VRMHumanoidRig extend VRMHumanoid

`vrm.humanoid.rig.humanoid` was too weird

0b5vr committed:
refactor: yarn lint-fix

0b5vr committed:
fix: VRMHumanBoneParentMap, forgot to rename thumb bones

Thumb bones has been renamed in https://github.com/pixiv/three-vrm/pull/1001

0b5vr committed:
fix: `VRMHumanoid.rig` makes a circular dependency, removing it temporarily

ke456-png committed:
HumanoidRigをHumanoidのメンバとして実装

Humanoidは正規化されていないModelRigと、正規化されているHumanoidRigの二つを持つように実装する.

HumanoidRigを操作することでモデルごとのローカル座標系を意識せずに関節を回転できる.

HumanoidRigへの操作はUpdateごとにModelRigへ反映させる

HumanoidRigのUpdateはデフォルトではVRMCoreのUpdateと同時に呼ばれ、autoUpdateHumanoidで設定する.

ke456-png committed:
HumanoidのインターフェースはHumanoidRigを参照するように変更

VRMの姿勢を変更する際には正規化済みのHumanoidRigの方が使いやすいことが多い.

よって基本的にはAutoUpdateを有効にし、Humanoidのインターフェースから姿勢を操作することを想定している.

ke456-png committed:
Mixamoのアニメーションの流し込みのexample

Rootモーションについて考慮漏れがあり修正する必要がある.
- MUST: HipsとRootの位置はアニメーションで変更できる必要がある
- MUST: Hipsの親は原点とは限らず、回転やスケールが入っている可能性がある

ke456-png committed:
MixamoアニメーションのexampleをCoreの側にも追加

HumanoidRigのデバッグのためにMixamoアニメーションのexampleを複製した.
VRM差し替えや0.x <-> 1.0の回転を考えるとVRMUtilsを使用したい箇所がある.

ke456-png committed:
ルートアニメーション用にHipsの位置を同期させる処理を追加

ModelRigのHipsのワールド座標をHumanoidRigのHipsへ追従させる処理を追加

VRMの全体のルート（VRM.Scene）とHipsはアニメーションさせるときに座標を移動させることが多いため対応する.

ke456-png reviewed (commented):
ルートアニメーション用のHipsの追従処理を実装

ke456-png commented on packages/three-vrm-core/src/humanoid/VRMHumanoidRig.ts:
```diff
@@ -124,6 +124,17 @@ export class VRMHumanoidRig extends VRMRig {
           .multiply(parentWorldRotation)
           .premultiply(invParentWorldRotation)
           .multiply(boneRotation);
+
+        // ルートアニメーション用
+        if (boneName === 'hips') {
+          const boneWorldPosition = rigBoneNode.getWorldPosition(new THREE.Vector3());
+          const parentWorldMatrix = boneNode.parent?.matrixWorld;
+
+          if (parentWorldMatrix) {
+            const localPosition = boneWorldPosition.applyMatrix4(parentWorldMatrix);
+            boneNode.position.copy(localPosition);
+          }
+        }
```
Hipsの追従処理

ke456-png committed:
ModelRigのHipsの親はvrm.scene以下のものが指定されている

modelRigで参照してるnodeはgltf.scene下のBoneが渡されているはず

ke456-png committed:
Helperの見た目上の問題を修正

examplesでHelperがモデルよりも上に描画されるように変更

モデルを切り替えたときに、前のモデルのHelperを削除する

ke456-png committed:
簡易的なコメントを追加

ke456-png committed:
Mixamoアニメーションのexampleへの動線を追加

0b5vr commented on packages/three-vrm-core/examples/humanoid.html:
```diff
@@ -108,9 +108,11 @@
+					humanoid.humanoidRig.getBoneNode( 'rightUpperArm' ).rotation.x = s;
 
+
+					currentGltf.userData.vrmCore.update(deltaTime);
```
[MAY] 文章の進め方としては、

```js
const vrmCore = currentGltf.userData.vrmCore;
const humanoid = vrmCore.humanoid;
```

としてから話し始める方が個人的には好きです。

ke456-png replied:
はい、そちらの方が良さそうなので修正しました。

https://github.com/pixiv/three-vrm/pull/889/commits/608812a01ec78063f2cc934e5f803f64c941120f#diff-a0463531f2cd444c3e6cc443ad79e4578772a8201be3a154aacfe42ce09e3d3a

0b5vr commented on packages/three-vrm-core/examples/humanoid.html:
```diff
@@ -57,7 +57,7 @@
 			loader.register( ( parser ) => {
 
-				return new THREE_VRM_CORE.VRMCoreLoaderPlugin( parser, { helperRoot: scene } );
+				return new THREE_VRM_CORE.VRMCoreLoaderPlugin( parser, { helperRoot: scene ,autoUpdateHumanoid:true} );
```
[MUST] lint

ke456-png replied:
修正しました。
https://github.com/pixiv/three-vrm/pull/889/commits/48d8f669844a1b08e5ea8dd5bef7de3da7f073fa

0b5vr commented on packages/three-vrm-core/src/VRMCoreLoaderPlugin.ts:
```diff
@@ -25,10 +25,11 @@ export class VRMCoreLoaderPlugin implements GLTFLoaderPlugin {
     this.parser = parser;
 
     const helperRoot = options?.helperRoot;
+    const autoUpdate = options?.autoUpdateHumanoid;
```
[SHOULD] `autoUpdateHumanoid` は主語デカすぎ・対象が曖昧すぎないかな……？
実際update対象となっているのはHumanoidRigのみですよね。

0b5vr commented on packages/three-vrm-core/src/VRMCoreLoaderPlugin.ts:
```diff
@@ -25,10 +25,11 @@ export class VRMCoreLoaderPlugin implements GLTFLoaderPlugin {
     this.expressionPlugin = options?.expressionPlugin ?? new VRMExpressionLoaderPlugin(parser);
     this.firstPersonPlugin = options?.firstPersonPlugin ?? new VRMFirstPersonLoaderPlugin(parser);
-    this.humanoidPlugin = options?.humanoidPlugin ?? new VRMHumanoidLoaderPlugin(parser, { helperRoot });
+    this.humanoidPlugin = options?.humanoidPlugin ?? new VRMHumanoidLoaderPlugin(parser, { helperRoot, autoUpdate });
```
[SHOULD] こちらの `autoUpdate` も同様。
両方まとめて、 `autoUpdateHumanoidRig` と呼んでしまって良いんじゃないかなと思っています。

ke456-png replied:
autoUpdateHumanoidRigに統一しました

https://github.com/pixiv/three-vrm/pull/889/commits/608812a01ec78063f2cc934e5f803f64c941120f

0b5vr commented on packages/three-vrm-core/src/humanoid/VRMRig.ts:
```diff
@@ -0,0 +1,192 @@
+/**
+ * A class represents the Rig of a VRM.
+ */
+export class VRMRig {
```
これって実はむしろHumanoidのベースクラスとして機能していたりしませんか？

ke456-png replied:
はい、現状ではVRMRigはHumanoidのベースクラスに近いものとして機能しています。

しかし、意図としてはVRMRigはHumanoidのベースクラスではありません。

VRMRigは0.xのHumanoidをベースとしています。
そして、1.0のHumanoidは0.xのHumanoidのように振舞えるようにpublicメソッドを用意してあるため
VRMRigはHumanoidは近いものになっています。

0b5vr replied:
internalなクラスになった上で、一旦 `VRMHumanBonesController` みたいなお名前にしてみませんか

0b5vr replied:
Normalizedのほうが `VRMNormalizedHumandBonesController` とかになるのかな

0b5vr commented:
なるほど、mixamoのexampleはthree-vrm・three-vrm-core双方に用意してくれたのですね。良さそうです

ke456-png committed:
fix lint

ke456-png committed:
AutoUpdateHumanoidの変数名の変更

すべてAutoUpdateHumanoidRigに統一しました。

examplesでの変数宣言での指摘箇所を修正

0b5vr commented on packages/three-vrm-core/examples/humanoid.html:
```diff
@@ -103,14 +103,17 @@
+					humanoid.humanoidRig.getBoneNode( 'rightUpperArm' ).rotation.x = s;
 
+
+					currentGltf.userData.vrmCore.update(deltaTime);
```
```suggestion
					vrmCore.update( deltaTime );
```

こっちをこう書けるようになるのですよ

ke456-png replied:
失念していました、修正しました。
https://github.com/pixiv/three-vrm/pull/889/commits/9860488a804646e8e5d01101a9ed245fcd565bdf

ke456-png commented on packages/three-vrm-core/src/humanoid/VRMHumanoid.ts:
```diff
@@ -24,13 +22,30 @@ export class VRMHumanoid {
    */
   public restPose: VRMPose;
 
+  /**
+   * A raw rig of the VRM.
+   */
+  public modelRig: VRMRig;
+
```
VRMHumanoid.humanBonesを持っている必要は無さそう

ke456-png committed:
宣言した変数を利用するようにexamplesを修正

ke456-png commented:
口頭：

今回このMRでは「正規化されたHumanBonesを操作するためのインタフェース」を実装します。

以下の方針で実装を行います。
- HumanoidRig / ModelRig / VRMRigはprivateとして隠蔽して、getBone類だけを公開します。
  - 内部だけで管理し、外部に公開しないシグネチャ命名の重要性を下げるためです。


- Humanoidに対しアクセサの追加・名前変更を行います.
  - get/set
    - `getRawBoneNode()`
    - `getNormalizedBone()`
    - …
  - `.normalizedHumanBones`
    - `.nomalizedHumanBones.humanBones`

- Humnoid.getBone()等は削除します
  - 返すボーンが正規化されているか自明でないためです。
  - getBoneで正規化したボーンを返し、従来の0.xの挙動を再現することも検討しましたが、今回は採用しません。


 HumanoidRig / ModelRig / VRMRig という命名は最適ではないため、名称の変更を検討していますが、
それは内部的な変更にとどまるようにします。

ke456-png committed:
Humanoidは生のボーンと正規化されたボーンへの参照手段を提供する

正規化されたBonesなどはprivateとして隠蔽して、getBone類だけを公開します。
内部だけで管理し、外部に公開しないシグネチャ命名の重要性を下げるためです。

autoUpdateHumanoidRigフラグおよびUpdateでnormalizedBoneの姿勢でrawBonesを上書きします。

Humnoid.getBone()等は返すボーンが正規化されているか自明でないため削除します。

ke456-png committed:
改名に伴う修正

ke456-png committed:
逆行列にし忘れていたので修正

ke456-png committed:
HumanoidのAutoUpdateとLookAtが競合する問題の修正

- LookAtとHumanoidのUpdateの実行順を変更
    - Humanoidを一番最初に実行し、上書き出来るようにする。

- LookAt内で正規化・非正規化ボーンの両方に対して回転を行う

0b5vr commented on packages/three-vrm-core/examples/humanoid.html:
```diff
@@ -57,7 +57,7 @@
 			loader.register( ( parser ) => {
 
-				return new THREE_VRM_CORE.VRMCoreLoaderPlugin( parser, { helperRoot: scene } );
+				return new THREE_VRM_CORE.VRMCoreLoaderPlugin( parser, { helperRoot: scene, autoUpdateHumanBones: true } );
```
Question: `autoUpdateHumanBones` はデフォルト `true` ですか？これを明示的に指定するのを推奨したいですか？

ke456-png replied:
> autoUpdateHumanBones はデフォルト true ですか？

はい、`autoUpdateHumanBones` はデフォルト `true` にしています。
three.js のautoUpdate系の変数はデフォルトで`true`だった記憶があるのでそちらに合わせた形です。

> これを明示的に指定するのを推奨したいですか？

いいえ、実際に利用する際にはわざわざ指定する必要はないと考えています。

しかし、humanoid・humanoidAnimationのexampleでは明示的に指定した方が良いと考えています。
このexampleではautoUpdateHumanBonesに依存しているので明示的に指定して情報を増やしたいという意図がありました。

0b5vr committed:
docs: fix doc comments of VRMHumanoid

- the name of argument `humanBones` was outdated
- add `autoUpdateHumanBones`

0b5vr committed:
docs: add a doc comment to `VRMHumanoidLoaderPluginOptions.autoUpdateHumanBones`

0b5vr committed:
change: don't expose `VRMHumanoidRig`

0b5vr committed:
change: expose VRMHumanoidLoaderPluginOptions

0b5vr committed:
refactor: rename private members of VRMHumanoid

- `_rawHumanBonesRig` -> `_rawHumanBones`
- `_normalizedHumanBonesRig` -> `_normalizedHumanBones`

0b5vr committed:
change: VRMHumanoid.copy should also copy `.autoUpdateHumanBones`

0b5vr committed:
change (1.0, humanoid): `VRMHumanoid`, add `rawRestPose` and `normalizedRestPose`
deprecate `restPose`

0b5vr committed:
refactor (1.0): VRMHumanoid, reorder methods

accessors should be placed before the constructor

Is there any eslint rule to force this? (I'm pretty sure there is)

0b5vr committed:
docs: update doc comments

0b5vr committed:
change: add a method `rawHumanBones`

It's too confusing to not having this I think

0b5vr committed:
change (1.0): VRMHumanoid, add deprecated compat methods

0b5vr committed:
refactor: lint-fix

0b5vr commented:
@ke456-png warningが出るdeprecated compatを追加してみました。良さそうですかね？
https://github.com/pixiv/three-vrm/pull/889/commits/5ea3c0de01549482d667c7efe146cc0b8daafce4

0b5vr committed:
refactor: mixamo animation example is now Mr.doob's Code Style™ compliant

0b5vr committed:
examples: index page, add description on humanoidAnimation

ke456-png committed:
fix: autoUpdateHumanBonesをLoaderOptionに追加

optionにautoUpdateHumanBonesを追加後、VRMLoaderで追加漏れがあったため対応

0b5vr commented:
https://github.com/pixiv/three-vrm/pull/889/commits/3e365fff1baed78d9bdd3831b07b18458e40c5cb : 確かに、これは見落としてました！ありがとうございます。

0b5vr commented:
normalized human bonesのルートは `VRMHumanoid` からアクセスできたほうが良いかもな。 `VRMHumanoid.normalizedHumanBoneRoot` とかですか

および、これはVRMのSceneに追加しちゃっても良い気がします。Expressionsと同じノリで

ke456-png committed:
読み込み時にnormalizedBoneのRootをgltf.sceneに追加する

アニメーション再生時などnormalizedBoneのRootがシーンに追加されている必要があるため

またHumanoidにnormalizedHumanBonesRootを参照するためのgetterを追加

0b5vr commented:
mixamoのモーションの仕様変更がここ数日？に入ったらしく、モデルがうまくアニメーションしなくなっちゃった。が、とりあえずこのブランチはこのままマージします。

![image](https://user-images.githubusercontent.com/7824814/187406683-8a213dc8-fe98-4d4e-8d8f-07afad914399.png)
